analyzers/tests/SonarAnalyzer.UnitTest/TestCases/InfiniteRecursion.RoslynCfg.cs

```diff
@@ -0,0 +1,504 @@
+                    }
+                };
+            }
+        }
+
+        MethodExpressionDelegate delMain = delegate {
+            void Foo()  // Noncompliant
+            {
+                Foo();
+            }
+        };
+
+        private void LamdaInIf(List<string> input)
+        {
+            if (input.Where(x =>
+                            {
+                                return Foo();
+                                bool Foo() { return Foo(); }  // Noncompliant
+                            }).SingleOrDefault() != null)
+            {
+            }
+        }
+    }
+
+    class MoreCases
+    {
+        void CallsOnThis()    // Noncompliant
+        {
+            this.CallsOnThis();
+        }
+
+        void CallsOnObject()  // Noncompliant
+        {
+            var x = new MoreCases();
+            x.CallsOnObject();
+        }
+
+        public virtual void CallsOnObjectVirtual(MoreCases arg)  // Noncompliant
+        {
+            arg.CallsOnObjectVirtual(arg);
+        }
+
+        static int PassItself(int a)
+        {
+            Receiver(PassItself);
+            return 42;
+        }
+
+        static void Receiver(Func<int, int> func)
+        {
+
+        }
+    }
+
+    class Exceptions
+    {
+        public int Property1 => throw new NotImplementedException();
+
+        public int Property2
+        {
+            get
+            {
+                throw new NotImplementedException();
+            }
+
+            set
+            {
+                throw new NotImplementedException();
+            }
+        }
+
+        public int Property3
+        {
+            get  // Noncompliant
+            {
+                var a = Property3;
+                throw new NotImplementedException();
+            }
+            set  // Noncompliant
+            {
+                Property3 = 42;
+                throw new NotImplementedException();
+            }
+        }
+
+        void ThrowsException() => throw new AggregateException();
+
+        void ThrowsExceptionInIf()
+        {
+            var a = 5;
+            if (a > 1)
+            {
+                throw new AggregateException();
+            }
+            else
+            {
+                throw new AggregateException();
+            }
+        }
+
+        void CallsItselfAndThrows()            // Noncompliant
+        {
+            CallsItselfAndThrows();
+            throw new AggregateException();
+        }
+
+        void ThrowsExceptionInIfAfterGoto()    // Noncompliant
+        {
+            A:
+            goto A;
+            var a = 5;
+            if (a > 1)
+            {
+                throw new AggregateException();
+            }
+        }
+
+        int ThrowsInIf(int a)
+        {
+            if (a > 1)
+            {
+                throw new ArgumentException();
+            }
+            A:
+            goto A;
+
+            return 42;
+        }
+
+        int ThrowsInTry(int a)
```

Comment:
Let's add something like this, just to document the case because finally blocks are also special.
```suggestion
        int ThrowsInFinally(int a)
        {
        	try
        	{
        	}
        	finally
        	{
        		throw new ArgumentException();
        	}
        }
```